server/admin/tree_gc.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package admin
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/google/trillian"
+	"github.com/google/trillian/storage"
+)
+
+var (
+	timeNow   = time.Now
+	timeSleep = time.Sleep
+)
+
+// DeletedTreeGC garbage collects deleted trees.
+//
+// Tree deletion goes through two separate stages:
+// * Soft deletion, which "flips a bit" (see Tree.Deleted) but otherwise leaves the tree unaltered
+// * Hard deletion, which effectively removes all tree data
+//
+// DeletedTreeGC performs the transition from soft to hard deletion. Trees that have been deleted
+// for at least DeletedThreshold are eligible for garbage collection.
+type DeletedTreeGC struct {
+	// Admin is the storage.AdminStorage interface.
+	Admin storage.AdminStorage
+
+	// DeleteThreshold defines the minimum time a tree has to remain in the soft-deleted state
+	// before its eligible for garbage collection.
+	DeleteThreshold time.Duration
+
+	// MinRunInterval defines how frequently sweeps for deleted trees are performed.
+	// Actual runs happen randomly between [minInterval,2*minInterval).
+	MinRunInterval time.Duration
+}
+
+// Run starts the tree garbage collection process. It never returns.
```

Comment:
Shouldn't it return if ctx is cancelled?

Reply:
Done. Good call.